sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -185,8 +183,21 @@ case class DataSource(
     }
 
     SchemaUtils.checkColumnNameDuplication(
-      (dataSchema ++ partitionSchema).map(_.name), "in the data schema and the partition schema",
-      sparkSession.sessionState.conf.caseSensitiveAnalysis)
+      dataSchema.map(_.name), "in the data schema", equality)
+    SchemaUtils.checkColumnNameDuplication(
+      partitionSchema.map(_.name), "in the partition schema", equality)
+
+    // We just print a waring message if the data schema and partition schema have the duplicate
+    // columns. This is because we allow users to do so in the previous Spark releases and
+    // we have the existing tests for the cases (e.g., `ParquetHadoopFsRelationSuite`).
+    // See SPARK-18108 and SPARK-21144 for related discussions.
+    try {
+      SchemaUtils.checkColumnNameDuplication(
```

Comment:
Since `dataSchema` and `partitionSchema` is firstly fixed by this function (`getOrInferFileFormatSchema`), IIUC we easily couldn't put this check in the constructer. If we put there, we need to move some code to decide schemas inside this function there. Thought?